### PR TITLE
Stable structures update

### PR DESCRIPTION
**Pull request by F3kilo**
Stable structures uses less memory and provide a more convenient API.

### CODE REVIEW

**Review by itsyaasir — CHANGES_REQUESTED** [review]
Everything LGTM, Just a note though, this is not part of your PR, but it seems one of the packages fails to build the `cmc-mock`, the CI didn't pick that because, it has not been added as a dep, but do check out.

**Comment by F3kilo** [comment]
> Everything LGTM, Just a note though, this is not part of your PR, but it seems one of the packages fails to build the `cmc-mock`, the CI didn't pick that because, it has not been added as a dep, but do check out.

Fixed

**Review by itsyaasir — APPROVED** [review]

File: ic-stable-structures/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "ic-stable-structures"
-version = "0.2.0"
+version = "0.4.0"
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/Cargo.toml:3]: It was there!)
Looks like some merge from `main` reset version back from `0.3` to `0.2` some time ago...

File: ic-stable-structures/src/multimap.rs [hunk 2]
```diff
@@ -150,71 +140,119 @@ struct KeyPair<K1, K2> {
...
     }
 
     pub fn first_key(&self) -> K1 {
         let offset = Self::size_prefix_len();
-        K1::from_bytes(self.encoded[offset..offset + self.first_key_len].to_vec())
+        K1::from_bytes(self.encoded[offset..offset + self.first_key_len].into())
     }
 
-    pub fn first_key_prefix(first_key: &K1) -> Result<Vec<u8>> {
+    pub fn min_key(first_key: &K1) -> Self {
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/src/multimap.rs:206]: It is minimal KeyPair with the specified `first_key`. I'll place a comment there.

File: ic-factory/src/state.rs [hunk 3]
```diff
@@ -52,16 +53,14 @@ impl Storable for CanisterHash {
         self.0.as_slice().into()
     }
 
-    fn from_bytes(bytes: Vec<u8>) -> Self {
-        Self(bytes)
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-factory/src/state.rs:56]: Done

File: ic-factory/src/state.rs [hunk 4]
```diff
@@ -108,7 +107,7 @@ impl Storable for StorableCanisterModule {
             .into()
     }
 
-    fn from_bytes(bytes: Vec<u8>) -> Self {
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-factory/src/state.rs:110]: Done

File: ic-factory/src/state.rs [hunk 5]
```diff
@@ -563,7 +553,7 @@ impl Storable for FactoryConfiguration {
...
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-factory/src/state.rs:556]: Done

File: ic-factory/src/state.rs [hunk 6]
```diff
@@ -655,23 +645,24 @@ async fn send_remaining_fee_to(
...
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
 struct PrincipalKey(Principal);
 
 impl Storable for PrincipalKey {
     fn to_bytes(&self) -> std::borrow::Cow<'_, [u8]> {
         self.0.as_slice().into()
     }
 
-    fn from_bytes(bytes: Vec<u8>) -> Self {
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-factory/src/state.rs:656]: Done

File: ic-stable-structures/src/multimap.rs [hunk 7]
```diff
@@ -150,71 +140,119 @@ struct KeyPair<K1, K2> {
...
         let second_key_bytes = second_key.to_bytes();
 
-        if first_key_bytes.len() > K1::max_size() as usize {
-            return Err(Error::ValueTooLarge(first_key_bytes.len() as _));
-        }
-
-        if second_key_bytes.len() > K2::max_size() as usize {
-            return Err(Error::ValueTooLarge(second_key_bytes.len() as _));
-        }
+        assert!(first_key_bytes.len() <= K1::MAX_SIZE as usize);
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/src/multimap.rs:185]: I've created a ticket for that.

File: ic-stable-structures/src/multimap.rs [hunk 8]
```diff
@@ -238,11 +276,11 @@ where
         Cow::Borrowed(&self.encoded)
     }
 
-    fn from_bytes(bytes: Vec<u8>) -> Self {
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/src/multimap.rs:279]: Done

File: ic-stable-structures/src/multimap.rs [hunk 9]
```diff
@@ -278,15 +316,14 @@ impl<V> Storable for Value<V> {
         Cow::Borrowed(&self.0)
     }
 
-    fn from_bytes(bytes: Vec<u8>) -> Self {
-        Self(bytes, PhantomData)
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/src/multimap.rs:319]: Done

File: ic-stable-structures/src/multimap.rs [hunk 10]
```diff
@@ -401,30 +438,29 @@ mod test {
             Cow::Owned(self.0.to_vec())
         }
 
-        fn from_bytes(bytes: Vec<u8>) -> Self {
+        fn from_bytes(bytes: Cow<[u8]>) -> Self {
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/src/multimap.rs:441]: Done

File: ic-stable-structures/src/test_utils.rs [hunk 11]
```diff
@@ -11,15 +13,13 @@ impl Storable for StringValue {
         self.0.to_bytes()
     }
 
-    fn from_bytes(bytes: Vec<u8>) -> Self {
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/src/test_utils.rs:16]: Done

File: ic-stable-structures/src/unbounded.rs [hunk 12]
```diff
@@ -58,51 +61,49 @@ where
...
     }
 
     /// Add or replace a value associated with `key`.
-    pub fn insert(&mut self, key: &K, value: &V) -> Result<()> {
-        let mut inner_key = Key::new(key)?;
+    ///
+    /// # Preconditions:
+    ///   - `key.to_bytes().len() <= K::MAX_SIZE`
+    pub fn insert(&mut self, key: &K, value: &V) -> Option<V> {
+        let mut inner_key = Key::new(key);
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/src/unbounded.rs:72]: Done

File: ic-stable-structures/src/unbounded.rs [hunk 13]
```diff
@@ -268,18 +286,17 @@ impl<K: BoundedStorable> Storable for Key<K> {
         (&self.data).into()
     }
 
-    fn from_bytes(bytes: Vec<u8>) -> Self {
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/src/unbounded.rs:289]: Done

File: ic-stable-structures/src/unbounded.rs [hunk 14]
```diff
@@ -310,23 +327,22 @@ impl<V: SlicedStorable> Storable for Chunk<V> {
         (&self.chunk).into()
     }
 
-    fn from_bytes(bytes: Vec<u8>) -> Self {
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/src/unbounded.rs:330]: Done

File: ic-stable-structures/src/multimap.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (F3kilo) [reply ic-stable-structures/src/multimap.rs:185]: https://infinityswap.atlassian.net/browse/CPROD-1512

**Pull request merged**